// @million/lint/dist/runtime/index.js
"use strict";var mt=Object.create;var q=Object.defineProperty;var ut=Object.getOwnPropertyDescriptor;var dt=Object.getOwnPropertyNames;var ft=Object.getPrototypeOf,gt=Object.prototype.hasOwnProperty;var yt=(e,t)=>{for(var n in t)q(e,n,{get:t[n],enumerable:!0})},fe=(e,t,n,o)=>{if(t&&typeof t=="object"||typeof t=="function")for(let r of dt(t))!gt.call(e,r)&&r!==n&&q(e,r,{get:()=>t[r],enumerable:!(o=ut(t,r))||o.enumerable});return e};var M=(e,t,n)=>(n=e!=null?mt(ft(e)):{},fe(t||!e||!e.__esModule?q(n,"default",{value:e,enumerable:!0}):n,e)),bt=e=>fe(q({},"__esModule",{value:!0}),e);var Ft={};yt(Ft,{captureClassSetState:()=>Ze,captureDeps:()=>Qe,captureFunctionCall:()=>et,captureFunctionDef:()=>tt,captureHTTP:()=>st,captureJSX:()=>nt,captureProps:()=>ot,captureReducer:()=>it,captureRef:()=>rt,invalidate:()=>J,registerComponent:()=>ct,registerFile:()=>lt,reportOutliers:()=>W,resync:()=>G,useMount:()=>at});module.exports=bt(Ft);var te={data:{},React:{},sync:[],embed:{},cwd:"",network:{keys:[],requests:[]},meta:{version:null,framework:null},invalidate:!0,active:!0,_lookup:{component:new WeakMap,hash:new Map}},g=()=>te,$=te._lookup.component,X=e=>!e||!$.has(e)?null:$.get(e),K=te._lookup.hash,k=(e,t)=>{if(!K.has(e))return null;let{filename:n,componentName:o,reference:r,prev:i}=K.get(e);return{filename:n,componentName:o,reference:r,prev:i[t],setPrev:l=>i[t]=l}},V=(...e)=>{!1&&console.log(...e)};var J=async()=>{let e=g();await G();for(let t=0,n=e.sync.length;t<n;t++)St(t)},G=async()=>{let e=g();for(let t in e.data){let n=e.data[t];for(let o in n.components){let r=n.components[o];r.renders=[]}}for(let t=0,n=e.network.requests.length;t<n;t++){let o=e.network.requests[t];o.totalTime=0}e.sync=await ne()},St=async e=>{let n=g().sync[e];try{let r=(await fetch(`http://localhost:${n.port}/invalidate`)).body?.getReader(),i=new TextDecoder;for(;r;){let{done:l,value:s}=await r.read();if(l)break;try{let a=Number(i.decode(s).replace("data:","").trim());n.version!==a&&(n.version=a,G())}catch{V("[Million Lint] Error parsing update")}}}catch(o){if(o?.message?.indexOf("network error")===-1)return;setTimeout(()=>{J()},1e3)}},ne=async(e=52921)=>{if(e>52963)throw new Error("[Million Lint] Could not find running VSCode! Did you install the extension?");let t=g(),n=await fetch(`http://localhost:${e}`).catch(()=>null);if(!n?.ok)return ne(e+1);let o=await n.json(),r=[];if(!Array.isArray(o))return ne(e+1);for(let i=0,l=o.length;i<l;i++){let s=o[i];if(!s.cwd)continue;let a=!1;for(let c in t.data)if(c.includes(s.cwd)){a=!0;break}a&&(isNaN(s.version)||r.push(s))}return r};var ie=e=>e.sort((t,n)=>t-n),be=e=>e.reduce((t,n)=>t+n,0),re=e=>be(e)/e.length,ye=e=>{let t=re(e),n=e.map(o=>(o-t)**2);return Math.sqrt(be(n)/(e.length-1))},xe=e=>{let t=e.filter(c=>c>0),n=ie(t),o=re(n),r=ye(n),i=n.filter(c=>c<o+2*r),l=re(i),s=ye(i),a=i.filter(c=>c<l+s);return Math.max(...a)};var W=()=>{let e=g(),t=Rt(),{all:n,outliers:o}=Tt(t);e.embed.render?.(o);for(let r=0,i=e.sync.length;r<i;r++){let l=e.sync[r];vt(l,o,n),wt(l)}},Rt=()=>{let e=g(),t=[],n=[],o=[];for(let r in e.data){let i=e.data[r].components;for(let l in i){let s=i[l];s.filename=r,t.push(s);for(let a=s.renders.length-1;a>=0;a--){let c=s.renders[a];n.push(c.selfTime*s.instances),o.push(c.count)}}}return{components:t,selfTimes:n,counts:o}},Tt=e=>{let{components:t,selfTimes:n,counts:o}=e,r=ie(n),i=xe(r),l=[],s=[];for(let a=0,c=t.length;a<c;a++){let p=t[a],m=!1,d=0,u=0,S=p.instances,y=0,h=0,R=p.instances;for(let f=0,E=p.renders.length;f<E;f++){let b=p.renders[f],w=["props","state","value"].includes(b.kind);d+=b.totalTime,u+=b.selfTime,w&&(S+=b.count),!(b.selfTime*p.instances<=i)&&(m=!0,y+=b.totalTime,h+=b.selfTime,w&&(R+=b.count))}p.loc&&(s.push({instances:p.instances,componentName:p.componentName,filename:p.filename,loc:p.loc,totalTime:d,selfTime:u,count:S}),m&&l.push({instances:p.instances,componentName:p.componentName,filename:p.filename,loc:p.loc,totalTime:y,selfTime:h,count:R}))}return{all:s,outliers:l}},vt=(e,t,n)=>{let o=g();o.invalidate=!1;try{he(e,{runtime:{version:e.version||0,outliers:t,all:n,data:o.data,meta:{anyaVersion:"0.0.66",version:o.meta.version,framework:o.meta.framework}}}),V("[Million Lint] Sending data to your VSCode editor",t)}catch(r){V("[Million Lint] Post report error",r)}},wt=e=>{let t=g();t.invalidate=!1;let n=t.network.requests;n.sort((r,i)=>i.totalTime-r.totalTime);let o=[];for(let r=0,i=n.length;r<i&&!(r<i*.25);r++){let l=n[r];l.totalTime<1||o.push(l)}try{he(e,{network:{all:n,outliers:o}}),V("[Million Lint] Sending network info to your VSCode editor",t.network)}catch(r){V("[Million Lint] Post report error",r)}},he=async(e,t)=>{try{await fetch(`http://localhost:${e.port}/?react-dev-overlay-crash-check`,{method:"GET",mode:"no-cors"});let o=await(await fetch(`http://localhost:${e.port}/ingest?t=${Object.keys(t).join(",")}`,{body:JSON.stringify(t),method:"POST",headers:{"Content-Type":"application/json"}})).json();if(!o.ok){let r=o.version;e.version=r,G(),V("[Million Lint] Error",o,r)}}catch{G()}};var Se=()=>{let e=g();setTimeout(()=>{J()},1e3),setInterval(()=>{!e.active||!e.invalidate||W()},1e3)};var Y=require("@twind/core"),Fe=M(require("react")),Ie=require("react-dom/client"),dn=require("construct-style-sheets-polyfill");var Re=require("@twind/core"),Te=M(require("@twind/preset-tailwind")),ve=M(require("@twind/preset-autoprefix")),we=(0,Re.defineConfig)({presets:[(0,ve.default)(),(0,Te.default)()]});var B=require("cmdk"),T=M(require("react")),Ce=require("zustand"),Oe=M(require("pretty-ms"));var _=M(require("react"));var ke=({className:e=""})=>_.default.createElement("svg",{width:46,height:32,viewBox:"0 0 68 48",fill:"none",xmlns:"http://www.w3.org/2000/svg",style:{width:"1.5rem"},className:`border-none ${e}`},_.default.createElement("path",{fillRule:"evenodd",clipRule:"evenodd",d:"M17.2639 40.5003C20.3923 36.048 26.7587 33.0001 34.1037 33.0001C41.4485 33.0001 47.8151 36.048 50.9435 40.5003C47.8151 44.9522 41.4485 48 34.1037 48C26.7587 48 20.3923 44.9522 17.2639 40.5003Z",fill:"white"}),_.default.createElement("path",{d:"M17.1488 7.42031C20.2753 3.01552 26.6376 0 33.9778 0C41.318 0 47.6806 3.01552 50.8069 7.42031C47.6806 11.8251 41.318 14.8406 33.9778 14.8406C26.6376 14.8406 20.2753 11.8251 17.1488 7.42031Z",fill:"white"}),_.default.createElement("path",{d:"M34.3012 24.0885C29.7526 28.4565 23.5834 30.9104 17.1506 30.9104C10.7179 30.9104 4.54862 28.4565 0 24.0885L17.1506 7.42093L34.3012 24.0885Z",fill:"white"}),_.default.createElement("path",{d:"M68 24.0966C63.4513 28.4647 57.2818 30.9186 50.8494 30.9186C44.4166 30.9186 38.2473 28.4647 33.6986 24.0966L50.8494 7.42093L68 24.0966Z",fill:"white"})),Ee=()=>_.default.createElement("svg",{className:`${A("h-5 w-5 text-white mx-[6px]")} animate-spinner`,xmlns:"http://www.w3.org/2000/svg",fill:"none",viewBox:"0 0 24 24"},_.default.createElement("circle",{style:{opacity:.25},cx:"12",cy:"12",r:"10",stroke:"currentColor",strokeWidth:"4"}),_.default.createElement("path",{style:{opacity:.75},fill:"currentColor",d:"M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"})),Ne=()=>_.default.createElement("svg",{className:"pt-[1px]",xmlns:"http://www.w3.org/2000/svg",width:"1em",height:"1em",viewBox:"0 0 24 24"},_.default.createElement("g",{fill:"none",stroke:"currentColor"},_.default.createElement("path",{d:"M20.5 6h-17m15.333 2.5l-.46 6.9c-.177 2.654-.265 3.981-1.13 4.79-.865.81-2.195.81-4.856.81h-.774c-2.66 0-3.99 0-4.856-.81-.865-.809-.953-2.136-1.13-4.79l-.46-6.9"}),_.default.createElement("path",{d:"M9.5 11l.5 5m4.5-5l-.5 5M6.5 6h.11a2 2 0 001.83-1.32l.034-.103.097-.291c.083-.249.125-.373.18-.479a1.5 1.5 0 011.094-.788C9.962 3 10.093 3 10.355 3h3.29c.262 0 .393 0 .51.019a1.5 1.5 0 011.094.788c.055.106.097.23.18.479l.097.291A2 2 0 0017.5 6",opacity:.5})));var se=(0,Ce.create)(e=>({open:!1,setOpen:t=>{e({open:t})}})),_e=({outliers:e})=>{let t=g(),n=se(s=>s.setOpen),[o,r]=(0,T.useState)("linear"),i=(0,T.useRef)(null),l=(0,T.useRef)(null);return(0,T.useEffect)(()=>{i?.current?.focus();let s=a=>{a.key==="Escape"&&n(!1)};return window.addEventListener("keydown",s),()=>{window.removeEventListener("keydown",s)}},[]),T.default.createElement("div",{className:"million-lint"},T.default.createElement(B.Command,{value:o,onValueChange:s=>{r(s)},loop:!0},T.default.createElement(B.Command.Input,{ref:i,autoFocus:!0,placeholder:"Search..."}),T.default.createElement("hr",{"cmdk-million-lint-loader":""}),T.default.createElement(B.Command.List,{ref:l,style:{colorScheme:"dark"}},T.default.createElement(B.Command.Empty,null,"No results found."),e.length>0&&T.default.createElement(B.Command.Group,{heading:"Issues"},e.sort((s,a)=>a.selfTime*a.instances-s.selfTime*s.instances).map(s=>{let{componentName:a,filename:c,loc:p}=s,[m,d]=p,u=a+c+m+d,S=()=>{window.open(`vscode://file/${c}:${m}:${d}`),n(!1)};return T.default.createElement(B.Command.Item,{value:a,onSelect:S,key:u,className:"py-5",title:c.replace(t.cwd,"")},T.default.createElement("span",{className:"font-semibold"},s.componentName)," ",T.default.createElement("span",{className:"mr-auto opacity-60 text-sm"},"\xD7",s.count,"\xA0(",(0,Oe.default)(s.selfTime,{compact:!0}),")"),T.default.createElement("span",{"cmdk-million-lint-meta":""},T.default.createElement("div",{className:"inline-block items-center justify-center rounded-[8px] text-gray-900 bg-white font-bold text-sm px-[8px] py-[3px] mx-1"},T.default.createElement("div",{className:"flex items-center justify-center gap-0.5"},"Open \u21B5"))))})),T.default.createElement(B.Command.Group,{heading:"Commands"},T.default.createElement(B.Command.Item,{value:"Reset and scan for new data",onSelect:()=>{n(!1),G()}},T.default.createElement(Ne,null),"Reset")))))},Ae=`
.million-lint [cmdk-root] {
  width: 100%;
  background: var(--gray1);
  border-radius: 12px;
  padding: 8px 0;
  min-width: 640px;
  font-family: var(--font-sans);
  box-shadow: var(--cmdk-shadow);
  border: 1px solid var(--gray6);
  position: relative;
  outline: none;
}
.million-lint [cmdk-root] kbd {
  font-family: var(--font-sans);
  background: var(--gray3);
  color: var(--gray11);
  height: 20px;
  width: 20px;
  border-radius: 4px;
  padding: 0 4px;
  display: flex;
  align-items: center;
  justify-content: center;
}
.million-lint [cmdk-root] kbd:first-of-type {
  margin-left: 8px;
}
.million-lint [cmdk-input] {
  font-family: var(--font-sans);
  border: none;
  width: 100%;
  font-size: 15px;
  padding: 8px 16px;
  outline: none;
  background: var(--bg);
  color: var(--gray12);
}
.million-lint [cmdk-input]::placeholder {
  color: var(--gray9);
}
.million-lint [cmdk-million-lint-loader] {
  --loader-color: var(--gray9);
  border: 0;
  width: 100%;
  width: 100%;
  left: 0;
  height: 1px;
  background: var(--gray6);
  position: relative;
  overflow: visible;
  display: block;
  margin-top: 12px;
  margin-bottom: 12px;
}

.million-lint [cmdk-item] {
  content-visibility: auto;
  cursor: pointer;
  height: 40px;
  border-radius: 8px;
  font-size: 14px;
  display: flex;
  align-items: center;
  gap: 8px;
  padding: 0 8px;
  color: var(--gray12);
  user-select: none;
  will-change: background, color;
  transition: all 150ms ease;
  transition-property: none;
}
.million-lint [cmdk-item][data-selected='true'] {
  background: var(--gray4);
  color: var(--gray12);
}
.million-lint [cmdk-item][data-disabled='true'] {
  color: var(--gray8);
  cursor: not-allowed;
}
.million-lint [cmdk-item]:active {
  transition-property: background;
  background: var(--gray4);
}
.million-lint [cmdk-item]:first-child {
  margin-top: 8px;
}
.million-lint [cmdk-item] + [cmdk-item] {
  margin-top: 4px;
}
.million-lint [cmdk-item] svg {
  width: 18px;
  height: 18px;
}
.million-lint [cmdk-million-lint-meta] {
  color: var(--gray11);
  font-size: 13px;
}
.million-lint [cmdk-list] {
  padding: 0 8px;
  height: 393px;
  overflow: auto;
  overscroll-behavior: contain;
  scroll-padding-block-end: 40px;
  transition: 100ms ease;
  transition-property: height;
}
.million-lint [cmdk-million-lint-open-trigger], .million-lint [cmdk-million-lint-subcommand-trigger] {
  color: var(--gray11);
  padding: 0px 4px 0px 8px;
  border-radius: 6px;
  font-weight: 500;
  font-size: 12px;
  height: 28px;
  letter-spacing: -0.25px;
}
.million-lint [cmdk-million-lint-clipboard-icon], .million-lint [cmdk-million-lint-hammer-icon] {
  width: 20px;
  height: 20px;
  border-radius: 6px;
  display: flex;
  align-items: center;
  justify-content: center;
  color: #fff;
}
.million-lint [cmdk-million-lint-clipboard-icon] svg, .million-lint [cmdk-million-lint-hammer-icon] svg {
  width: 14px;
  height: 14px;
}
.million-lint [cmdk-million-lint-clipboard-icon] {
  background: linear-gradient(to bottom, #f55354, #eb4646);
}
.million-lint [cmdk-million-lint-hammer-icon] {
  background: linear-gradient(to bottom, #6cb9a3, #2c6459);
}
.million-lint [cmdk-million-lint-open-trigger] {
  display: flex;
  align-items: center;
  color: var(--gray12);
}
.million-lint [cmdk-million-lint-subcommand-trigger] {
  display: flex;
  align-items: center;
  gap: 4px;
  right: 8px;
  bottom: 8px;
}
.million-lint [cmdk-million-lint-subcommand-trigger] svg {
  width: 14px;
  height: 14px;
}
.million-lint [cmdk-million-lint-subcommand-trigger] hr {
  height: 100%;
  background: var(--gray6);
  border: 0;
  width: 1px;
}
.million-lint [cmdk-million-lint-subcommand-trigger][aria-expanded='true'], .million-lint [cmdk-million-lint-subcommand-trigger]:hover {
  background: var(--gray4);
}
.million-lint [cmdk-million-lint-subcommand-trigger][aria-expanded='true'] kbd, .million-lint [cmdk-million-lint-subcommand-trigger]:hover kbd {
  background: var(--gray7);
}
.million-lint [cmdk-separator] {
  height: 1px;
  width: 100%;
  background: var(--gray5);
  margin: 4px 0;
}
.million-lint *:not([hidden]) + [cmdk-group] {
  margin-top: 8px;
}
.million-lint [cmdk-group-heading] {
  user-select: none;
  font-size: 12px;
  color: var(--gray11);
  padding: 0 8px;
  display: flex;
  align-items: center;
}
.million-lint [cmdk-million-lint-footer] {
  display: flex;
  height: 40px;
  align-items: center;
  width: 100%;
  position: absolute;
  background: var(--gray1);
  bottom: 0;
  padding: 8px;
  border-top: 1px solid var(--gray6);
  border-radius: 0 0 12px 12px;
}
.million-lint [cmdk-million-lint-footer] svg {
  width: 20px;
  height: 20px;
  filter: grayscale(1);
  margin-right: auto;
}
.million-lint [cmdk-million-lint-footer] hr {
  height: 12px;
  width: 1px;
  border: 0;
  background: var(--gray6);
  margin: 0 4px 0px 12px;
}
.million-lint [cmdk-million-lint-footer] {
    background: var(--gray2);
 }
.million-lint [cmdk-dialog] {
  position: fixed;
  left: 50%;
  top: var(--page-top);
  transform: translateX(-50%);
}
.million-lint [cmdk-dialog] [cmdk] {
  transform-origin: center center;
  animation: dialogIn var(--transition-fast) forwards;
}
.million-lint [cmdk-dialog][data-state='closed'] [cmdk] {
  animation: dialogOut var(--transition-fast) forwards;
}
.million-lint [cmdk-empty] {
  font-size: 14px;
  display: flex;
  align-items: center;
  justify-content: center;
  height: 64px;
  white-space: pre-wrap;
  color: var(--gray11);
}
[cmdk-raycast-footer] {
  display: flex;
  height: 40px;
  align-items: center;
  width: 100%;
  position: absolute;
  background: var(--gray2);
  bottom: 0;
  padding: 8px;
  border-top: 1px solid var(--gray6);
  border-radius: 0 0 12px 12px;
}

[cmdk-raycast-footer] hr {
    height: 12px;
    width: 1px;
    border: 0;
    background: var(--gray6);
    margin: 0 4px 0px 12px;
}

[cmdk-raycast-open-trigger] {
  display: flex;
  align-items: center;
  color: var(--gray12);
}

[cmdk-raycast-open-trigger],
[cmdk-raycast-subcommand-trigger] {
  color: var(--gray11);
  padding: 0px 4px 0px 8px;
  border-radius: 6px;
  font-weight: 500;
  font-size: 12px;
  height: 28px;
  letter-spacing: -0.25px;
}

@keyframes slideIn {
  0% {
    opacity: 0;
    transform: scale(0.96);
 }
  100% {
    opacity: 1;
    transform: scale(1);
 }
}
@media (max-width: 640px) {
  .million-lint [cmdk-root] {
    min-width: 100%;
  }
  .million-lint [cmdk-input] {
    font-size: 16px;
  }
}
.variables {
  --font-sans: -apple-system, BlinkMacSystemFont, avenir next, avenir, segoe ui, helvetica neue, helvetica, Cantarell, Ubuntu, roboto, noto, arial, sans-serif;
  --app-bg: var(--gray1);
  --lowContrast: #000;
  --highContrast: #fff;
  --gray1: #161616;
  --gray2: #1c1c1c;
  --gray3: #232323;
  --gray4: #282828;
  --gray5: #2e2e2e;
  --gray6: #343434;
  --gray7: #3e3e3e;
  --gray8: #505050;
  --gray9: #707070;
  --gray10: #7e7e7e;
  --gray11: #a0a0a0;
  --gray12: #ededed;
  --grayA1: rgba(255, 255, 255, 0);
  --grayA2: rgba(255, 255, 255, 0.026);
  --grayA3: rgba(255, 255, 255, 0.056);
  --grayA4: rgba(255, 255, 255, 0.077);
  --grayA5: rgba(255, 255, 255, 0.103);
  --grayA6: rgba(255, 255, 255, 0.129);
  --grayA7: rgba(255, 255, 255, 0.172);
  --grayA8: rgba(255, 255, 255, 0.249);
  --grayA9: rgba(255, 255, 255, 0.386);
  --grayA10: rgba(255, 255, 255, 0.446);
  --grayA11: rgba(255, 255, 255, 0.592);
  --grayA12: rgba(255, 255, 255, 0.923);
}
`;var j=M(require("@radix-ui/react-dialog")),P=M(require("@radix-ui/react-tooltip")),x=M(require("react")),Le=require("react-start-transition"),Me=M(require("react-draggable"));var Pe=({root:e,disabled:t})=>{let n=g(),[o,r]=se(p=>[p.open,p.setOpen]),[i,l]=(0,x.useState)([]),[s,a]=(0,Le.useTransition)(),c=i.length&&n.sync.length;return n.embed.render=p=>{a(()=>{l(p)})},t?null:x.default.createElement(P.Provider,{delayDuration:300},x.default.createElement(P.Root,null,x.default.createElement("div",{className:`million-embed ${c?"active":""}`},x.default.createElement(j.Root,{open:o,onOpenChange:p=>{c&&r(p)},modal:!1},x.default.createElement(Me.default,{disabled:!c},x.default.createElement(P.Trigger,{asChild:!0},x.default.createElement("div",{className:A("flex touch-action-none cursor-pointer py-1 px-2 rounded-full bg-[#18181b] border border-gray-800 shadow-md backdrop-blur-50  overflow-hidden")},x.default.createElement("div",{className:A(`flex items-center justify-center gap-1 ${o?"opacity-50":""}`)},x.default.createElement(j.Trigger,{asChild:!0},x.default.createElement("a",{className:A("flex items-center justify-center w-8 h-8 rounded-full bg-[#18181b]"),target:"_blank"},x.default.createElement(ke,null))),x.default.createElement("div",{className:A("h-[1.25rem] w-0.5 bg-[#313131]")}),x.default.createElement(j.Trigger,{asChild:!0},x.default.createElement("div",null,c?x.default.createElement(Et,{outliers:i.length}):x.default.createElement(Ee,null))))))),c?x.default.createElement(j.Portal,{container:e},x.default.createElement("div",{className:`fixed inset-0 bg-black opacity-70 transition-all ${o?"":"hidden pointer-events-none"}`,onClick:()=>{r(!1)}}),x.default.createElement("div",{className:`${A("z-[1000] fixed top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2")} variables dark`},x.default.createElement(_e,{outliers:i}))):x.default.createElement(P.Portal,{container:e},x.default.createElement(P.Content,{className:"TooltipContent",sideOffset:5},n.sync.length?"Scanning for slow components...":"Connecting to Visual Studio Code..."))))))},Et=({outliers:e})=>x.default.createElement("div",{className:"flex relative"},x.default.createElement("div",{className:A("relative flex items-center justify-center rounded-full bg-gradient-to-t from-[#c29ce9] to-[#ab8ae7] text-transparent font-semibold text-sm px-[8px] py-[2px] mx-1 animate-notification")},e),x.default.createElement("div",{className:A("absolute flex items-center justify-center rounded-full bg-gradient-to-t from-[#c29ce9] to-[#ab8ae7] text-white font-semibold text-sm px-[8px] py-[2px] mx-1")},e)),De=`
.million-embed {
  position: fixed;
  bottom: -35px;
  z-index: 2147483647;
  transition: all 0.3s;
  left: 50%;
  transform: translate(-50%, 0%);
  opacity: 0.5;
}
.million-embed:hover {
  opacity: 1;
  bottom: 5px;
}
.active {
  bottom: 5px;
  opacity: 1;
}

.TooltipContent {
  border-radius: 10px;
  padding: 10px 15px;
  font-size: 15px;
  line-height: 1;
  color: white;
  background-color: #313131;
  box-shadow: hsl(206 22% 7% / 35%) 0px 10px 38px -10px, hsl(206 22% 7% / 20%) 0px 10px 20px -15px;
  user-select: none;
  animation-duration: 400ms;
  animation-timing-function: cubic-bezier(0.16, 1, 0.3, 1);
  will-change: transform, opacity;
}

@keyframes scaleIn {
  from {
    opacity: 0;
    transform: scale(0);
  }
  to {
    opacity: 1;
    transform: scale(1);
  }
}

@keyframes spinner {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
.animate-spinner {
  animation: spinner 1s cubic-bezier(.96,.31,.37,.92) infinite;
}

@keyframes notification {
  75%, 100% {
    transform: scale(1.4);
    opacity: 0;
  }
}
.animate-notification {
  animation: notification 1s cubic-bezier(0, 0, 0.2, 1) infinite;
}
`;var Ge=(0,Y.cssom)(new CSSStyleSheet),A=(0,Y.twind)(we,Ge),Be=()=>{let e=document.createElement("div");e.style.all="revert";let t=e.attachShadow({mode:"open"}),n=document.createElement("style"),o=document.createElement("div");t.adoptedStyleSheets=[Ge.target],(0,Y.observe)(A,t),document.body.appendChild(e),t.appendChild(n),t.appendChild(o);let r=`
    ${De}
    ${Ae}
  `;n.textContent=r,(0,Ie.createRoot)(o).render(Fe.default.createElement(Pe,{root:o}))};var Ue=()=>{let e=g();typeof window>"u"||(window.__ANYA_GLOBAL_DATA__=e,Se(),Be(),document.addEventListener("visibilitychange",()=>{e.active=document.visibilityState==="visible"}))};var Ve=e=>["mount","unmount","nested-update","update"].includes(e.kind);var ae=e=>e.kind==="call",le=e=>e.kind==="props";var Nt=(e,t)=>{let n=!0;for(let o=0,r=e.length;o<r;o++)if(e[o]!==t[o]){n=!1;break}return n},Ct=(e,t)=>{if(!(Nt(e.loc,t.loc)&&e.kind===t.kind&&e.owner===t.owner))return!1;for(let o=0,r=e.changes.length;o<r;o++){let i=e.changes[o],l=t.changes[o];if(!i||!l)continue;if(!(i.prev===l.prev&&i.next===l.next))return!1}return!0},Ot=(e,t)=>{if(!e&&!t)return[];if(!e?.length)return t;if(!t?.length)return e;let n=new Map;for(let o=0,r=e.length;o<r;o++)n.set(e[o].name,o);for(let o=0,r=t.length;o<r;o++){let i=t[o],l=n.get(i.name);if(l===void 0){let a=e.push(i);n.set(i.name,a);continue}let s=e[l];s.count++,s.next=i.next,s.prev=i.prev,s.unstable||(s.unstable=i.unstable)}return e},_t=(e,t)=>{if(!e&&!t)return[];if(!e?.length)return t;if(!t?.length)return e;let n=new Map;for(let o=0,r=e.length;o<r;o++)n.set(e[o].prev+e[o].next,o);for(let o=0,r=t.length;o<r;o++){let i=t[o],l=n.get(i.prev+i.next);if(l===void 0){let a=e.push(i);n.set(i.prev+i.next,a);continue}let s=e[l];s.count++,s.next=i.next,s.prev=i.prev,s.unstable||(s.unstable=i.unstable)}return e},At=(e,t)=>{if(!e&&!t)return[];if(!e?.length)return t;if(!t?.length)return e;let n=new Map;for(let o=0,r=e.length;o<r;o++)n.set(e[o].caller,o);for(let o=0,r=t.length;o<r;o++){let i=t[o],l=n.get(i.caller);if(l===void 0){let a=e.push(i);n.set(i.caller,a);continue}let s=e[l];s.count++,s.next=i.next,s.prev=i.prev,s.unstable||(s.unstable=i.unstable)}return e},N=(e,t,n)=>{let o=g();o.invalidate=!0;let r=o.data[e];if(!r)return;let i=r.components[t];if(Ve(n)&&n.self){let[s,a]=n.self.split("::"),c=o.data[s].components[a];a!==t&&(i.children[a]||(i.children[a]=n.self),c.parents[t]||(c.parents[t]=`${e}::${t}`))}if(!i)return;let l=!1;for(let s=i.renders.length-1;s>=0;s--){let a=i.renders[s];if(a&&Ct(a,n)){l=!0,a.count++,a.totalTime+=n.totalTime,a.selfTime+=n.selfTime,le(n)&&le(a)?a.changes=Ot(a.changes,n.changes):ae(n)&&ae(a)?a.changes=At(a.changes,n.changes):a.changes=_t(a.changes,n.changes);break}}(!l||!i.renders.length)&&i.renders.push(n)},je=(e,t,n,o,r,i,l,s,a,c)=>({kind:e,_loc:n,loc:t,selfTime:o,totalTime:r,count:i,filename:l,self:s,owner:a,changes:c}),U=(e,t,n,o,r,i,l,s)=>({kind:e,loc:t,selfTime:n,totalTime:o,count:r,filename:i,self:null,owner:l,changes:s}),$e=(e,t,n,o,r,i,l,s)=>({kind:e,loc:t,selfTime:n,totalTime:o,count:r,filename:i,self:null,owner:l,changes:s}),Ye=(e,t,n,o,r,i,l,s,a)=>({kind:e,loc:t,selfTime:n,totalTime:o,count:r,filename:i,owner:l,self:null,changes:s,locs:a});var ze="Symbol(react.context)";var ce="Symbol(react.forward_ref)";var pe="Symbol(react.fragment)";var me="Symbol(react.lazy)";var ue="Symbol(react.memo)";var Xe="Symbol(react.provider)";var C=["object","function"];var He=require("react");function qe(e,t,n){if(!e||t(e))return e;let o=n?e.return:e.child;for(;o;){let r=qe(o,t,n);if(r)return r;o=n?null:o.sibling}}function L(){let e=g(),t=e.React.useRef();return e.React.useState(()=>{let n=Function.prototype.bind;Function.prototype.bind=function(o,r){return o===null&&typeof r?.type=="function"&&(t.current=r,Function.prototype.bind=n),n.apply(this,arguments)}}),t.current}function Ke(e){let t=L(),n=e.useRef();return e.useLayoutEffect(()=>{n.current=qe(t,o=>typeof o.type=="string",!1)?.stateNode},[t]),n}var D=e=>{let t=e?.actualDuration??0,n=t,o=e?.child;for(;t>0&&o!=null;)n-=o?.actualDuration||0,o=o.sibling;return{totalTime:t,selfTime:n}},Je=new Map,Z=new Map,z=e=>{let t=Je.get(e);if(typeof t=="string")return t;let n=null;if(e&&(typeof e.displayName=="string"?n=e.displayName:typeof e.name=="string"&&e.name!==""&&(n=e.name)),!n)return null;if(Z.has(n)){let o=Z.get(n)||2;Z.set(n,o+1),n+=`\`${String(o)}`}else Z.set(n,2);return Je.set(e,n),n},We=e=>{if(e)switch(typeof e){case"string":return e;case"function":return z(e);case"symbol":if(String(e)===pe)return"";default:if(e.$$typeof)switch(String(e.$$typeof)){case pe:return"";case ue:case ce:case me:return z(e);case Xe:{let t=e._context||e.context;return z(t)}case ze:{let t=e._context||e.context;return z(t)}default:return String(e.$$typeof)}}return"Unknown"},O=()=>{let t=g().React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner.current;if(!t)return null;let n=t?._debugSource?.fileName,o=t?._debugOwner?.type;return!n||!o?null:X(o)||`${n}::${z(o)}`},de=e=>{if(!(0,He.isValidElement)(e))return null;if(typeof e.type=="function")return e.type;if(typeof e.type!="object")return null;if("$$typeof"in e.type)switch(String(e.type.$$typeof)){case ue:case ce:case me:return de(e.type)}return null},F=e=>{let t=e?._debugOwner?.type;return X(t)??z(t)};var{hasOwnProperty:Lt,toString:Mt}=Object.prototype,Pt=e=>typeof e=="object"&&e!==null&&e.constructor===Object,Dt=(e,t)=>e.length>t?`${e.slice(0,t)}\u2026`:e,v=e=>{switch(typeof e){case"boolean":case"undefined":case"number":case"bigint":case"symbol":return String(e);case"function":return e.toString();case"string":return Dt(e,20);case"object":if(e===null)return"null";if(Array.isArray(e))return e.length>0?"[\u2026]":"[]";if(typeof e.$$typeof=="symbol"&&String(e.$$typeof)==="Symbol(react.element)")return`<${We(e.type)}${Object.keys(e.props).length>0?" \u2026":""}/>`;if(Pt(e)){for(let n in e)if(Lt.call(e,n))return"{\u2026}";return"{}"}let t=Mt.call(e).slice(8,-1);if(t==="Object"){let n=Object.getPrototypeOf(e)?.constructor;if(typeof n=="function")return`${n.displayName||n.name||""}{\u2026}`}return`${t}{\u2026}`}};var Ze=(e,t,n,o)=>{let r=k(e,o);if(!r)return t;let{filename:i,componentName:l,prev:s,setPrev:a}=r;function c(...p){let m=p[0],d=L(),u=F(d)??O(),{selfTime:S,totalTime:y}=D(d);if(t.apply(this,p),Object.is(s,m))return m;let h=v(s),R=v(m),f=m?typeof m:typeof s,b=[{unstable:s!==m&&h===R&&C.includes(f),prev:h,next:R,type:f,count:1}],w=U("state",n,S,y,1,i,u,b);return N(i,l,w),a(m),m}return c};var Qe=(e,t,n,o,r)=>{if(t??=[],!Array.isArray(t))return t;let i=k(e,r);if(!i)return t;let{filename:l,componentName:s,prev:a,setPrev:c}=i,p=L(),m=F(p)??O(),{selfTime:d,totalTime:u}=D(p);if(a){let S=[];for(let y=0,h=t.length;y<h;y++){let R=a[y],f=t[y];if(Object.is(R,f))continue;let E=v(R),b=v(f),w=f?typeof f:typeof R,I=E===b&&C.includes(w);S.push({unstable:I,prev:E,next:b,type:w,count:1,index:y})}if(S.length){let y=Ye("deps",n,d,u,1,l,m,S,o);N(l,s,y)}}return c(t),t};var et=(e,t,n,o,r)=>{let i=k(e,o);if(!i)return t;let{filename:l,componentName:s,prev:a,setPrev:c}=i,p=O();if(a){if(Object.is(a,t))return t;let m=v(a),d=v(t),u=t?typeof t:typeof a,S=m===d&&C.includes(u),h=U("value",n,0,0,1,l,p,[{prev:m,next:d,caller:r,unstable:S,type:u,count:1}]);N(l,s,h)}return c(t),t},tt=(e,t,n,o)=>{if(typeof t!="function")return t;let r=k(e,o);if(!r)return t;let{filename:i,componentName:l,prev:s,setPrev:a}=r,c=O();return function(...m){let d=performance.now(),u=t(...m);if(s){if(Object.is(s,u))return u;let S=performance.now()-d,y=v(s),h=v(u),R=u?typeof u:typeof s,f=y===h&&C.includes(R),b=U("value",n,S,S,1,i,c,[{prev:y,next:h,unstable:f,type:R,count:1}]);N(i,l,b)}return a(u),u}};var Q=require("react");var nt=(e,t,n,o,r)=>{let i=k(e,r);if(!i)return t;let{filename:l,componentName:s,prev:a,setPrev:c}=i,p=O(),m=de(t),d=`${l}::${s}`,u=X(m);return m&&!u&&$.set(m,d),c(m),(0,Q.createElement)(Q.Profiler,{children:t,id:d,onRender:(S,y,h,R)=>{let f=[];if(a&&!Object.is(a,m)){let b=v(a),w=v(m),I=m?typeof m:typeof a,H=b===w&&C.includes(I);f.push({prev:b,next:w,unstable:H,type:I,count:1})}let E=je(y,o,n,h,R,1,l,u,p,f);N(l,s,E),c(m)}})};var ot=(e,t,n,o)=>{if(t??={},typeof t!="object")return t;let r=k(e,o);if(!r)return t;let{filename:i,componentName:l,setPrev:s}=r,a=r.prev;if(!a){let h={prevProps:null,prevPropsKeys:null};s(h),a=h}let c=L(),p=F(c)??O(),{selfTime:m,totalTime:d}=D(c),u=a.prevProps,S=a.prevPropsKeys??(u?Object.keys(u):[]),y=[];if(u){y=Object.keys(t);let h=new Set(S.concat(y)),R=[];for(let f of h){let E=u[f],b=t[f];if(Object.is(E,b))continue;let w=v(E),I=v(b),H=b?typeof b:typeof E,pt=w===I&&C.includes(H);R.push({name:f,unstable:pt,prev:w,next:I,type:H,count:1})}if(R.length){let f=$e("props",n,m,d,1,i,p,R);N(i,l,f)}}return s({prevProps:t,prevPropsKeys:y}),t};var rt=(e,t,n,o)=>{let r=k(e,o);if(!r)return t;let{filename:i,componentName:l,prev:s,setPrev:a}=r,c=L(),p=F(c)??O();if(s){if(Object.is(s,t.current))return t;let{selfTime:m,totalTime:d}=D(c),u=v(s),S=v(t.current),y=t.current?typeof t.current:typeof s,h=u===S&&C.includes(y),f=U("value",n,m,d,1,i,p,[{prev:u,next:S,unstable:h,type:y,count:1}]);N(i,l,f),a(t.current)}return t};var it=(e,t,n,o)=>{if(!Array.isArray(t))return t;let r=k(e,o);if(!r)return t;let{filename:i,componentName:l,prev:s,setPrev:a}=r,c=L(),p=F(c)??O(),m=t[1];a(t[0]);let d=u=>{if(m(u),Object.is(s,u))return;let{selfTime:S,totalTime:y}=D(c),h=v(s),R=v(u),f=u?typeof u:typeof s,b=[{unstable:h===R&&C.includes(f),prev:h,next:R,type:f,count:1}],w=U("state",n,S,y,1,i,p,b);N(i,l,w)};return t[1]=d,t};var st=(e,t,n,o,r)=>{let i=k(e,r);if(!i)return t;let{filename:l}=i,s=g(),{keys:a,requests:c}=s.network,p=n.join("::")+l;if(a.includes(p)||(a.push(p),s.invalidate=!0,c.push({loc:n,strings:o,filename:l,totalTime:0}),"__ANYA__"in window.fetch))return t;let m=window.fetch,d=async(...u)=>{let S=performance.now(),y=await m(u[0],u[1]),h=u[0],R=performance.now()-S;if(typeof h=="string")for(let f=0,E=c.length;f<E;f++){let b=c[f];for(let w=0,I=b.strings.length;w<I;w++)b.strings[w]===h&&(b.totalTime+=R)}return y};return d.__ANYA__=!0,window.fetch=d,t};var ee=require("react");var at=(e,t)=>{let n=g(),o=Ke(n.React);(0,ee.useEffect)(()=>{let r=n.data[e];if(!r)return;let i=r.components[t];if(!i)return;let l;return document.readyState==="complete"?i.isLazyLoadable=!0:o.current?.nodeType===Node.ELEMENT_NODE&&(l=new IntersectionObserver(()=>{i.isLazyLoadable=!1,l.disconnect()},{threshold:1,root:null,rootMargin:"0px"}),(0,ee.startTransition)(()=>{l.observe(o.current)})),i.instances++,()=>{l?.disconnect(),i.instances--}},[])};var lt=(e,t,n,o)=>{let r=g();if(r.invalidate=!0,e==="")throw new Error("Unable to add source for empty filename");r.cwd=t,r.React=n,r.meta.framework=o,r.meta.version=n.version,r.data[e]={components:{}}},ct=(e,t,n,o,r,i,l)=>{let s=g();s.invalidate=!0;let a=s.data[e];if(!a)return;let c=a.components[t];if(c){c.loc=i,c.externals=r;return}a.components[t]={componentName:t,loc:i,externals:r,renders:[],children:{},parents:{},instances:0,get count(){let p=0,m=this.renders;for(let d=0,u=m.length;d<u;d++)["mount","unmount","nested-update","update"].includes(m[d].kind)||(p+=m[d].count);return p}},$.set(o,`${e}::${t}`),K.set(n,{filename:e,componentName:t,reference:o,prev:new Array(l),setPrev:()=>{}})};(()=>{let e=!1;e||(Ue(),e=!0)})();
